checker/src/org/checkerframework/checker/lock/wholefiles/LockAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,261 @@
+        addInheritedAnnotation(AnnotationUtils.fromClass(elements,
+                ReleasesNoLocks.class));
+        addInheritedAnnotation(AnnotationUtils.fromClass(elements,
+                LockingFree.class));
+    }
+
+    @Override
+    public QualifierHierarchy createQualifierHierarchy(MultiGraphFactory factory) {
+        return new LockQualifierHierarchy(factory);
+    }
+
+    @Override
+    protected LockAnalysis createFlowAnalysis(List<Pair<VariableElement, CFValue>> fieldValues) {
+        return new LockAnalysis(checker, this, fieldValues);
+    }
+
+    @Override
+    public LockTransfer createFlowTransferFunction(CFAbstractAnalysis<CFValue, LockStore, LockTransfer> analysis) {
+        return new LockTransfer((LockAnalysis) analysis,(LockChecker)this.checker);
+    }
+
+    protected AnnotationMirror getDeclAnnotationNoAliases(Element elt,
+            Class<? extends Annotation> anno) {
+        String annoName = anno.getCanonicalName().intern();
+        return getDeclAnnotation(elt, annoName, false);
+    }
+
+    protected QualifierDefaults getQualifierDefaults() {
+        return defaults;
+    }
+
+    class LockQualifierHierarchy extends MultiGraphQualifierHierarchy {
+
+        public LockQualifierHierarchy(MultiGraphFactory f) {
+            super(f, LOCKHELD);
+        }
+
+        private boolean isGuardedBy(AnnotationMirror am) {
+            return AnnotationUtils.areSameIgnoringValues(am, GUARDEDBY) ||
+                   AnnotationUtils.areSameIgnoringValues(am, JAVAXGUARDEDBY) ||
+                   AnnotationUtils.areSameIgnoringValues(am, JCIPGUARDEDBY);
+        }
+
+        @Override
+        public boolean isSubtype(AnnotationMirror rhs, AnnotationMirror lhs) {
+
+            boolean lhsIsGuardedBy = isGuardedBy(lhs);
+            boolean rhsIsGuardedBy = isGuardedBy(rhs);
+
+            if (lhsIsGuardedBy && rhsIsGuardedBy) {
+                // Two @GuardedBy annotations are considered subtypes of each other if and only if their values match exactly.
+
+                List<String> lhsValues =
+                    AnnotationUtils.getElementValueArray(lhs, "value", String.class, true);
+                List<String> rhsValues =
+                    AnnotationUtils.getElementValueArray(rhs, "value", String.class, true);
+
+                return rhsValues.containsAll(lhsValues) && lhsValues.containsAll(rhsValues);
+            }
+
+            // Remove values from @GuardedBy annotations (and use the Checker Framework's GuardedBy annotation, not JCIP's or Javax's)
+            // for further subtype checking.
+
+            return super.isSubtype(rhsIsGuardedBy ? GUARDEDBY : rhs, lhsIsGuardedBy ? GUARDEDBY : lhs);
+        }
+    }
+
+    // Returns the list of expressions specified in the @Holding (or equivalent) annotation on a method.
+    // Returns an empty list if such an annotation is not present.
+    protected List<String> methodHolding(ExecutableElement element) {
```

Comment:
It would be nice if this was in a JavaDoc comment so the IDEs show it at call sites.